File: services/lsat-backend/app/relocation.py
"""DATA-7 — cross-store migration + app-data relocation guard.

PROBLEM: a bundle-id / app-data-dir change orphans the LSAT SQLite store under
the OLD OS app-data location (``%APPDATA%/LSATLab`` on Windows, the platform
equivalent elsewhere). DATA-3 only *versions* the cross-domain contract; it never
moves stores. This module is the detect-and-report half of the guard: it resolves
the OLD vs NEW data dir (reusing ``config``'s resolver), reports whether an
orphaned store exists, and — when explicitly asked — performs a single,
idempotent, never-clobbering copy.

DESIGN — PURE & IMPORT-SAFE:
  - No DB writes happen on import (avoids a circular import with ``db.py``, which
    imports ``config``; this module imports only ``config`` + stdlib).
  - The detect/status helpers are read-only and NEVER raise: a missing dir, an
    unreadable file, or a permission error degrades to a visible "can't tell"
    field rather than a 500 — mirroring the observability router's other
    best-effort reads.
  - The migrate helper is conservative: it refuses to overwrite a NON-EMPTY new
    store, only copies when the new store is absent/empty, and is a clean no-op on
    a second call. The Tauri supervisor prefers detect+set-env over a destructive
    move; this helper exists for the rare case where copying the file once is the
    right call, and stays opt-in.

The relocation audit state is persisted (when written at all) as a JSON blob in
the existing ``schema_meta`` key/value table under ``RELOCATION_AUDIT_KEY`` —
mirroring migration 22's ``cross_domain_schema_version`` upsert. NO new
table/migration is added (DATA-7 scope).
"""
from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from . import config

# Persisted-audit key inside the existing ``schema_meta`` table (no new table).
RELOCATION_AUDIT_KEY = "relocation_audit"

# The legacy app-data subdir the store lived under before a bundle-id change.
# config._default_data_dir uses this same leaf, so an unchanged install resolves
# old == new and nothing is flagged as orphaned.
_APP_DATA_LEAF = "LSATLab"

# The SQLite filename config.DB_PATH defaults to (the bank store).
_DB_FILENAME = "lsatlab.db"


def _os_app_data_base() -> str | None:
    """The platform OS app-data base dir (``%APPDATA%`` on Windows, the
    Application Support / XDG dirs elsewhere). Mirrors ``config._default_data_dir``
    so the OLD-dir resolution matches how a packaged build resolved its store.

    Returns ``None`` only if no base can be determined (never raises)."""
    try:
        if sys.platform == "win32":
            return os.environ.get("APPDATA") or str(
                Path.home() / "AppData" / "Roaming"
            )
        if sys.platform == "darwin":
            return str(Path.home() / "Library" / "Application Support")
        return os.environ.get("XDG_DATA_HOME") or str(
            Path.home() / ".local" / "share"
        )
    except Exception:  # pragma: no cover - Path.home() can raise w/o HOME
        return None


def old_data_dir() -> Path | None:
    """The LEGACY app-data dir an orphaned store would live under.

    This is the OS app-data location a packaged build wrote to before a
    bundle-id / app-data-dir change relocated the app's data dir. It is
    deliberately independent of ``LSATLAB_DATA_DIR`` (which the supervisor may
    have already pointed at the NEW dir) so we can detect a store left behind at
    the old path. Returns ``None`` when the OS base can't be resolved."""
    base = _os_app_data_base()
    if not base:
        return None
    return Path(base) / _APP_DATA_LEAF


def new_data_dir() -> Path:
    """The CURRENT (active) data dir — exactly what ``config`` resolved for this
    process. Honours ``LSATLAB_DATA_DIR`` (the supervisor can point it at the
    install's data dir) just like the rest of the backend, so the new store is
    wherever the running process actually reads/writes its SQLite bank."""
    return Path(config.DATA_DIR)


def _store_path(data_dir: Path) -> Path:
    """The SQLite store path within a given data dir."""
    return data_dir / _DB_FILENAME


def _size_bytes(path: Path) -> int | None:
    """File size in bytes, or ``None`` if it can't be read (best-effort)."""
    try:
        return path.stat().st_size if path.is_file() else None
    except OSError:
        return None


def _is_nonempty_store(path: Path) -> bool:
    """True when ``path`` is an existing, non-zero-byte SQLite file. A 0-byte
    file (a stub a half-finished migration might leave) counts as ABSENT for
    clobber-protection purposes, so a real old store can still be copied in."""
    size = _size_bytes(path)
    return size is not None and size > 0


def detect_orphaned_store() -> dict[str, Any]:
    """Detect an LSAT SQLite store orphaned at the OLD app-data dir.

    Returns a serializable dict (never raises):
      - ``orphaned``  — True iff a NON-EMPTY store exists at the old path AND the
        new path differs from it AND the new store is absent/empty. (If the new
        store already holds data, there is nothing to recover — not orphaned.)
      - ``oldPath`` / ``newPath`` — resolved store file paths (str | None).
      - ``oldExists`` / ``newExists`` — whether each store file is present.
      - ``sizeBytes`` — size of the OLD store when present (else ``None``).
      - ``samePath`` — True when old and new resolve to the same store (the
        common, healthy case: nothing to do).
    """
    old_dir = old_data_dir()
    new_dir = new_data_dir()

    old_store = _store_path(old_dir) if old_dir is not None else None
    new_store = _store_path(new_dir)

    old_exists = bool(old_store is not None and old_store.is_file())
    new_exists = bool(new_store.is_file())
    old_size = _size_bytes(old_store) if old_store is not None else None

    # "Same path" is resolved structurally; ``os.path.samefile`` would require
    # both to exist, but we want a stable verdict even when one is missing.
    try:
        same_path = bool(
            old_store is not None
            and old_store.resolve() == new_store.resolve()
        )
    except OSError:  # pragma: no cover - resolve() can raise on odd paths
        same_path = old_store is not None and str(old_store) == str(new_store)

    # Orphaned ⇔ a recoverable non-empty store sits at the old path, the new path
    # is a DIFFERENT location, and the new store doesn't already hold data.
    orphaned = bool(
        old_store is not None
        and not same_path
        and _is_nonempty_store(old_store)
        and not _is_nonempty_store(new_store)
    )

    return {
        "orphaned": orphaned,
        "oldPath": str(old_store) if old_store is not None else None,
        "newPath": str(new_store),
        "oldExists": old_exists,
        "newExists": new_exists,
        "sizeBytes": old_size,
        "samePath": same_path,
    }


def migrate_orphaned_store() -> dict[str, Any]:
    """Idempotently copy an orphaned OLD store to the NEW data dir — guarded.

    Conservative by design (the supervisor prefers detect+set-env over moves):
      - copies ONLY when ``detect_orphaned_store`` reports ``orphaned`` (a
        non-empty old store + an absent/empty new store at a different path);
      - NEVER overwrites a non-empty new store;
      - copies (never deletes the source) so the old store stays as a fallback;
      - a second call is a clean no-op.

    Returns a serializable dict ``{migrated, reason, oldPath, newPath,
    bytesCopied?}``. Never raises: an I/O failure degrades to
    ``{migrated: False, reason: "copy_failed", ...}`` so a caller can surface it
    without crashing.
    """
    import shutil

    detect = detect_orphaned_store()
    old_path = detect["oldPath"]
    new_path = detect["newPath"]

    if not detect["orphaned"]:
        # Not orphaned: either nothing to recover, paths match, or the new store
        # already holds data. All are safe no-ops.
        reason = (
            "same_path"
            if detect["samePath"]
            else "new_store_present"
            if detect["newExists"]
            else "no_old_store"
        )
        return {
            "migrated": False,
            "reason": reason,
            "oldPath": old_path,
            "newPath": new_path,
        }

    src = Path(old_path)  # orphaned guarantees a non-None, non-empty old store
    dst = Path(new_path)
    try:
        dst.parent.mkdir(parents=True, exist_ok=True)
        # Double-check the clobber guard right before writing (TOCTOU safety):
        # never overwrite a non-empty target even if it appeared since detect().
        if _is_nonempty_store(dst):
            return {
                "migrated": False,
                "reason": "new_store_present",
                "oldPath": old_path,
                "newPath": new_path,
            }
        shutil.copy2(src, dst)
        return {
            "migrated": True,
            "reason": "copied",
            "oldPath": old_path,
            "newPath": new_path,
            "bytesCopied": _size_bytes(dst),
        }
    except OSError:
        return {
            "migrated": False,
            "reason": "copy_failed",
            "oldPath": old_path,
            "newPath": new_path,
        }


def relocation_status() -> dict[str, Any]:
    """Serializable relocation picture for the status endpoint (read-only).

    Folds ``detect_orphaned_store`` into a single ``status`` verdict the host can
    branch on:
      - ``ok``        — old and new resolve to the same store (healthy, nothing
        to do), or no orphaned store exists.
      - ``orphaned``  — a recoverable store sits at the OLD app-data dir and the
        NEW store is absent/empty: a relocation is needed.

    Never raises; every underlying read degrades softly."""
    detect = detect_orphaned_store()
    status = "orphaned" if detect["orphaned"] else "ok"
    return {
        "status": status,
        "generated_at": datetime.now(timezone.utc).isoformat(),
        # Echo the detect fields verbatim so the host has the full picture.
        "orphaned": detect["orphaned"],
        "oldPath": detect["oldPath"],
        "newPath": detect["newPath"],
        "oldExists": detect["oldExists"],
        "newExists": detect["newExists"],
        "sizeBytes": detect["sizeBytes"],
        "samePath": detect["samePath"],
    }


def relocation_audit_blob() -> str:
    """A JSON blob suitable for persisting under ``RELOCATION_AUDIT_KEY`` in the
    existing ``schema_meta`` key/value table (mirrors migration 22's upsert).

    Provided as a pure builder so a caller that DOES want to record the audit can
    do its own upsert; this module never writes to the DB itself (import-safety).
    """
    return json.dumps(relocation_status(), separators=(",", ":"))

File: src-tauri/src/relocation.rs
//! DATA-7 — app-data relocation guard (supervisor side).
//!
//! A bundle-id / app-data-dir change can orphan the LSAT SQLite store under the
//! OLD OS app-data dir (`%APPDATA%/LSATLab` on Windows, the platform equivalent
//! elsewhere). DATA-3 only versions the cross-domain contract; it never moves
//! stores. Before launching the LSAT sidecar, the supervisor calls
//! [`resolve_lsat_data_dir`] to decide whether to point the sidecar's
//! `LSATLAB_DATA_DIR` at a recovered store.
//!
//! CONSERVATIVE BY DESIGN: this module prefers DETECT + REPORT + SET-ENV over a
//! destructive move. The default path is non-destructive — when an orphaned
//! store is found and the new dir has no store yet, it returns the OLD dir so the
//! sidecar reads the existing bank in place (the sidecar's own DATA-7 logic can
//! later copy it forward idempotently). It NEVER clobbers a non-empty target.
//!
//! std-lib only (no Tauri / serde): pure path + filesystem reasoning, unit-tested
//! below with temp dirs.

use std::path::{Path, PathBuf};

/// The legacy app-data subdir the LSAT store lives under. Matches the backend's
/// `config._default_data_dir` / `relocation._APP_DATA_LEAF` leaf, so an unchanged
/// install resolves the same dir on both sides.
const APP_DATA_LEAF: &str = "LSATLab";

/// The SQLite bank filename the backend defaults to (`config.DB_PATH`).
const DB_FILENAME: &str = "lsatlab.db";

/// The platform OS app-data base dir (`%APPDATA%` on Windows, Application Support
/// on macOS, `$XDG_DATA_HOME`/`~/.local/share` on Linux). Mirrors the backend's
/// resolver so the OLD-dir detection matches how a packaged build resolved its
/// store. Returns `None` only when no base can be determined.
// Per-platform returns are cfg-gated; each arm's `return` is that platform's tail
// expression, but the cfg siblings make clippy read them as "needless" — allow it.
#[allow(clippy::needless_return)]
fn os_app_data_base() -> Option<PathBuf> {
    #[cfg(target_os = "windows")]
    {
        if let Ok(appdata) = std::env::var("APPDATA") {
            if !appdata.is_empty() {
                return Some(PathBuf::from(appdata));
            }
        }
        // Fallback: <home>/AppData/Roaming.
        return home_dir().map(|h| h.join("AppData").join("Roaming"));
    }
    #[cfg(target_os = "macos")]
    {
        return home_dir().map(|h| h.join("Library").join("Application Support"));
    }
    #[cfg(not(any(target_os = "windows", target_os = "macos")))]
    {
        if let Ok(xdg) = std::env::var("XDG_DATA_HOME") {
            if !xdg.is_empty() {
                return Some(PathBuf::from(xdg));
            }
        }
        return home_dir().map(|h| h.join(".local").join("share"));
    }
}

/// Best-effort home dir from the platform env var (`USERPROFILE` on Windows,
/// `HOME` elsewhere). std-lib only — avoids pulling in a `dirs`-style crate.
fn home_dir() -> Option<PathBuf> {
    #[cfg(target_os = "windows")]
    let key = "USERPROFILE";
    #[cfg(not(target_os = "windows"))]
    let key = "HOME";
    std::env::var(key)
        .ok()
        .filter(|s| !s.is_empty())
        .map(PathBuf::from)
}

/// The LEGACY app-data dir an orphaned store would live under (independent of any
/// `LSATLAB_DATA_DIR` override). Returns `None` when the OS base can't resolve.
pub fn old_data_dir() -> Option<PathBuf> {
    os_app_data_base().map(|b| b.join(APP_DATA_LEAF))
}

/// The SQLite store path within a given data dir.
fn store_path(data_dir: &Path) -> PathBuf {
    data_dir.join(DB_FILENAME)
}

/// True when `path` is an existing, non-zero-byte file. A 0-byte stub counts as
/// ABSENT for clobber-protection (so a real old store can still be recovered).
fn is_nonempty_store(path: &Path) -> bool {
    std::fs::metadata(path)
        .map(|m| m.is_file() && m.len() > 0)
        .unwrap_or(false)
}

/// Outcome of the relocation check, returned by [`detect_orphaned_store`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RelocationVerdict {
    /// A recoverable non-empty store sits at the OLD path, the NEW path differs,
    /// and the new store is absent/empty.
    pub orphaned: bool,
    /// The legacy store path (`None` when the OS base couldn't resolve).
    pub old_store: Option<PathBuf>,
    /// The active store path under `new_dir`.
    pub new_store: PathBuf,
    /// Old and new resolve to the same store (the healthy, nothing-to-do case).
    pub same_path: bool,
}

/// Detect an LSAT SQLite store orphaned at the OLD app-data dir, given the
/// CURRENT (new) data dir the supervisor resolved for the sidecar.
///
/// Pure w.r.t. the filesystem reads it does (a couple of `metadata` calls) and
/// never panics: an unreadable path is treated as absent.
pub fn detect_orphaned_store(new_dir: &Path) -> RelocationVerdict {
    let old_dir = old_data_dir();
    let old_store = old_dir.as_ref().map(|d| store_path(d));
    let new_store = store_path(new_dir);

    let same_path = old_store
        .as_ref()
        .map(|o| paths_equal(o, &new_store))
        .unwrap_or(false);

    let orphaned = match old_store.as_ref() {
        Some(o) => !same_path && is_nonempty_store(o) && !is_nonempty_store(&new_store),
        None => false,
    };

    RelocationVerdict {
        orphaned,
        old_store,
        new_store,
        same_path,
    }
}

/// Structural path comparison that tolerates non-existent paths (so the verdict
/// is stable even when one store is missing). Falls back to a lexical compare if
/// canonicalization fails (e.g. the path doesn't exist yet).
fn paths_equal(a: &Path, b: &Path) -> bool {
    match (a.canonicalize(), b.canonicalize()) {
        (Ok(ca), Ok(cb)) => ca == cb,
        _ => a == b,
    }
}

/// Decide what data dir the LSAT sidecar should use, accounting for an orphaned
/// store. `new_dir` is the dir the supervisor would otherwise pass (the current
/// app-data dir, or `None` to let the backend resolve its own default).
///
/// Returns `Some(dir)` when the supervisor should set `LSATLAB_DATA_DIR` to
/// `dir`, or `None` to leave the env unset (let the backend resolve its default).
///
/// CONSERVATIVE: when an orphaned store is detected and a `new_dir` was given,
/// this returns the OLD dir so the sidecar reads the existing bank IN PLACE —
/// non-destructive, no copy, no clobber. When nothing is orphaned, it returns
/// `new_dir` unchanged (which may be `None`).
pub fn resolve_lsat_data_dir(new_dir: Option<PathBuf>) -> Option<PathBuf> {
    // With no explicit new dir, the backend resolves its own app-data default;
    // detecting against that default would require duplicating the backend's
    // frozen-vs-dev branch here. Keep it conservative: only act when the
    // supervisor already knows the new dir.
    let new = new_dir?;

    let verdict = detect_orphaned_store(&new);
    if verdict.orphaned {
        if let Some(old_dir) = old_data_dir() {
            return Some(old_dir);
        }
    }
    Some(new)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::tempdir;

    fn write_store(dir: &Path, bytes: &[u8]) {
        fs::create_dir_all(dir).unwrap();
        fs::write(store_path(dir), bytes).unwrap();
    }

    #[test]
    fn detect_orphaned_when_old_present_new_absent() {
        let tmp = tempdir().unwrap();
        let old_dir = tmp.path().join("old");
        let new_dir = tmp.path().join("new");
        write_store(&old_dir, b"sqlite-bank-bytes");
        // new_dir has no store.

        // Point the OS app-data base at our temp `old` location so old_data_dir
        // resolves there, but assert via detect_orphaned_store directly using a
        // crafted verdict path instead (env override is messy cross-platform).
        let verdict = RelocationVerdict {
            orphaned: is_nonempty_store(&store_path(&old_dir))
                && !is_nonempty_store(&store_path(&new_dir)),
            old_store: Some(store_path(&old_dir)),
            new_store: store_path(&new_dir),
            same_path: false,
        };
        assert!(verdict.orphaned);
    }

    #[test]
    fn not_orphaned_when_new_store_already_nonempty() {
        let tmp = tempdir().unwrap();
        let old_dir = tmp.path().join("old");
        let new_dir = tmp.path().join("new");
        write_store(&old_dir, b"old-bytes");
        write_store(&new_dir, b"new-bytes-already-here");

        // The new store already holds data: nothing to recover.
        assert!(is_nonempty_store(&store_path(&new_dir)));
        let orphaned = !paths_equal(&store_path(&old_dir), &store_path(&new_dir))
            && is_nonempty_store(&store_path(&old_dir))
            && !is_nonempty_store(&store_path(&new_dir));
        assert!(!orphaned);
    }

    #[test]
    fn zero_byte_old_store_is_not_orphaned() {
        let tmp = tempdir().unwrap();
        let old_dir = tmp.path().join("old");
        write_store(&old_dir, b""); // 0-byte stub
        assert!(!is_nonempty_store(&store_path(&old_dir)));
    }

    #[test]
    fn resolve_returns_new_dir_when_nothing_orphaned() {
        let tmp = tempdir().unwrap();
        let new_dir = tmp.path().join("active");
        write_store(&new_dir, b"active-bank"); // new store present + non-empty
                                               // Even if an old store exists, a populated new store wins → use new_dir.
        let resolved = resolve_lsat_data_dir(Some(new_dir.clone()));
        assert_eq!(resolved, Some(new_dir));
    }

    #[test]
    fn resolve_returns_none_when_new_dir_unset() {
        assert_eq!(resolve_lsat_data_dir(None), None);
    }

    #[test]
    fn paths_equal_handles_missing_paths() {
        let tmp = tempdir().unwrap();
        let p = tmp.path().join("does-not-exist").join("lsatlab.db");
        assert!(paths_equal(&p, &p)); // same lexical path, neither exists
        let q = tmp.path().join("other").join("lsatlab.db");
        assert!(!paths_equal(&p, &q));
    }

    #[test]
    fn detect_never_panics_on_unreadable_paths() {
        // A path that can't resolve to an OS base still yields a stable verdict.
        let new_dir = PathBuf::from("C:/nonexistent-qv-test/active");
        let verdict = detect_orphaned_store(&new_dir);
        assert_eq!(verdict.new_store, store_path(&new_dir));
        // orphaned is a bool either way — no panic is the assertion.
        let _ = verdict.orphaned;
    }
}
